data_structure/concurrent_hash_map.h

```diff
@@ -0,0 +1,295 @@
+/**
+ * concurrent_hash_map.h
+ * 
+ * This implementation borrows from folly::ConcurrentHashMap the idea of sharding key space
```

Comment:
Why not use folly? Maybe this way will be easier for you to modify later during performance testing?

Reply:
Ah I forgot to mention. To use the map in folly, we need to make the whole folly library as a dependency, which is followed by a ton of other dependencies that folly needs and I want to avoid that. I've been scouting for a more stand-alone maps but they are either too simple, not suitable for our use case, or again part of a bigger library. 
In the old slog implementation, they used a single mutex to guard the whole map so I think we'll be fine with this one.